### PR TITLE
DM/mariadb: sync the gtid executed in slave during master-slave replication

### PR DESCRIPTION
**Pull request by okJiang**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tiflow/issues/10741

### What is changed and how it works?

- add test case
- update go-mysql version https://github.com/go-mysql-org/go-mysql/pull/852

#### Tests 

 - Unit test
 - Integration test 
 
~~https://github.com/okJiang/tiflow/actions/runs/8243030893~~
https://github.com/okJiang/tiflow/actions/runs/8290808978

### Release note 

```release-note
`None`
```

### CODE REVIEW

File: dm/tests/mariadb_master_down_and_up/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        - subnet: "172.28.128.0/24"
+
+# re-usable extension fields, https://docs.docker.com/compose/compose-file/#extension-fields.
+x-node: &default-node
+  privileged: true
+  restart: always
+  environment:
+    MYSQL_ROOT_PASSWORD: "123456"
+
+services:
```
Comment (okJiang) [inline dm/tests/mariadb_master_down_and_up/docker-compose.yml:16]:
```suggestion
# In this case, gtid_domain_id should be the same, and gtid_strict_mode should be OFF.
services:
```

**Review by GMHDBJD — COMMENTED** [review]
can you enable relay？rest LGTM

**Review by GMHDBJD — APPROVED** [review]
LGTM

File: dm/pkg/binlog/event/common.go [hunk 2]
```diff
@@ -207,7 +211,9 @@ func verifySingleGTID(flavor string, gSet gmysql.GTIDSet) (interface{}, error) {
 			return nil, terror.ErrBinlogOnlyOneGTIDSupport.Generate(len(mariaGTIDs.Sets), gSet)
 		}
 		var mariaGTID *gmysql.MariadbGTID
-		for _, mariaGTID = range mariaGTIDs.Sets {
+		for _, set := range mariaGTIDs.Sets {
+			for _, mariaGTID = range set {
```
Comment (lance6716) [inline dm/pkg/binlog/event/common.go:215]: check `len` is 1

File: dm/pkg/binlog/event/event.go [hunk 3]
```diff
@@ -736,21 +736,23 @@ func GenMariaDBGTIDListEvent(header *replication.EventHeader, latestPos uint32,
...
-		// Server_ID, 4 bytes
-		err = binary.Write(payload, binary.LittleEndian, mGTID.ServerID)
-		if err != nil {
-			return nil, terror.ErrBinlogWriteBinaryData.AnnotateDelegate(err, "write Server_ID %d", mGTID.ServerID)
-		}
-		// GTID sequence, 8 bytes
-		err = binary.Write(payload, binary.LittleEndian, mGTID.SequenceNumber)
-		if err != nil {
-			return nil, terror.ErrBinlogWriteBinaryData.AnnotateDelegate(err, "write GTID sequence %d", mGTID.SequenceNumber)
+	for _, set := range mariaDBGSet.Sets {
```
Comment (lance6716) [inline dm/pkg/binlog/event/event.go:739]: in line 733 there's `numOfGTIDs` which should also be updated. Before fix it, find out why the test didn't reveal the problem.

Comment (okJiang) [reply dm/pkg/binlog/event/event.go:739]: image
This function is used in test only 😅

Comment (okJiang) [reply dm/pkg/binlog/event/event.go:739]: Let me take a closer look at unit test

Comment (okJiang) [reply dm/pkg/binlog/event/event.go:739]: I found the reason why the unit test succeeds. All executions of GenMariaDBGTIDListEvent will execute
https://github.com/pingcap/tiflow/blob/6453fef89e71d3a76dd0b47f784c1e370dfb37e0/dm/pkg/binlog/event/util.go#L162 

This function will decode the event into the "correct" form to compare with other results. There is a high probability that other results will also be decoded through it.

I will add an unit test case for it.

Comment (okJiang) [reply dm/pkg/binlog/event/event.go:742]: https://github.com/pingcap/tiflow/pull/10753/commits/3b094a0358b60403809b17c1c814d50f29887819

File: dm/pkg/binlog/event/common.go (hunk 2 above)
Comment (lance6716) [reply dm/pkg/binlog/event/common.go:214]: we can check len(set) before the for loop

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: dm/pkg/binlog/event/common.go (hunk 2 above)
Comment (okJiang) [reply dm/pkg/binlog/event/common.go:214]: That can't get the right gtid numbers in the error message

**Review by lance6716 — APPROVED** [review]

**Review by lance6716 — APPROVED** [review]

**Pull request merged**